File: io/src/test/java/edu/ucr/cs/bdlab/beast/io/shapefile/DBFWriterTest.java
/*
 * Copyright 2018 University of California, Riverside
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.ucr.cs.bdlab.beast.io.shapefile;

import edu.ucr.cs.bdlab.beast.common.BeastOptions;
import edu.ucr.cs.bdlab.beast.geolite.Feature;
import edu.ucr.cs.bdlab.beast.geolite.IFeature;
import edu.ucr.cs.bdlab.test.JavaSpatialSparkTest;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.Path;
import org.apache.spark.sql.types.DataType;
import org.apache.spark.sql.types.DataTypes;
import scala.Tuple2;

import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Calendar;
import java.util.GregorianCalendar;

public class DBFWriterTest extends JavaSpatialSparkTest {

  public void testSimpleWriter() throws IOException {
    DBFReader dbfReader = new DBFReader();
    DBFWriter dbfWriter = new DBFWriter();
    String[] dbfFileNames = {"usa-major-cities"};
    for (String dbFileName : dbfFileNames) {
      Path dbfPath = new Path(scratchPath(), dbFileName+".dbf");
      Configuration conf = new Configuration();
      // Create a DBF file that is a copy of the existing file
      try {
        DataInputStream dbfInput = new DataInputStream(getClass().getResourceAsStream("/" + dbFileName + "/" + dbFileName + ".dbf"));
        dbfReader.initialize(dbfInput, new BeastOptions(conf));
        dbfWriter.initialize(dbfPath, conf);
        while (dbfReader.nextKeyValue()) {
          dbfWriter.write(dbfReader.getCurrentValue());
        }
      } finally {
        dbfReader.close();
        dbfWriter.close();
      }

      // Read the written file back and make sure it is similar to the original one
      DBFReader reader2 = new DBFReader();
      try {
        DataInputStream dbfInput = new DataInputStream(getClass().getResourceAsStream("/"+dbFileName+"/"+dbFileName+".dbf"));
        dbfReader.initialize(dbfInput, new BeastOptions(conf));
        reader2.initialize(new DataInputStream(new FileInputStream(dbfPath.toString())), new BeastOptions(conf));
        while (dbfReader.nextKeyValue()) {
          assertTrue("The written file contains too few records", reader2.nextKeyValue());
          assertEquals(dbfReader.getCurrentValue(), reader2.getCurrentValue());
        }
      } finally {
        dbfReader.close();
        reader2.close();
      }
    }
  }

  public void testShouldWriteFeaturesWithoutNames() throws IOException {
    // A sample feature
    Feature f = Feature.create(null, null, null, new Object[]{"abc", "def"});

    DBFWriter dbfWriter = new DBFWriter();
    Path dbfPath = new Path(scratchPath(), "test.dbf");
    Configuration conf = new Configuration();
    // Create a DBF file with one record
    try {
      dbfWriter.initialize(dbfPath, conf);
      dbfWriter.write(f);
    } finally {
      dbfWriter.close();
    }

    // Read the written file back and make sure it is correct
    try (DBFReader reader = new DBFReader()) {
      reader.initialize(dbfPath, new BeastOptions(conf));
      int count = 0;
      while (reader.nextKeyValue()) {
        count++;
        assertEquals(reader.getCurrentValue(), f);
      }
      assertEquals(1, count);
    }
  }

  public void testWriteAllAttributeTypes() throws IOException {
    DBFWriter dbfWriter = new DBFWriter();
    Path dbfPath = new Path(scratchPath(), "output.dbf");
    Configuration conf = new Configuration();
    // Create a DBF file with one feature that has all the possible types
    GregorianCalendar date = new GregorianCalendar(DBFConstants.UTC);
    date.clear();
    date.set(2019, Calendar.OCTOBER, 9);
    java.sql.Date dateValue = new java.sql.Date(date.getTimeInMillis());
    scala.collection.immutable.Map<String, String> tags = new scala.collection.immutable.HashMap<String, String>();
    tags = tags.$plus(new Tuple2<>("tag", "value"));
    Feature f = Feature.create(
        null,
        new String[]{"key1", "key2", "key3", "key4", "key5", "key6", "key7", "key8", "key9", "key10"},
        null,
        new Object[]{Boolean.TRUE, 123, (byte) 12, (short) 13, (long) 6546543, 12.25f, 12.655847, "value9", dateValue, tags}
    );
    try {
      dbfWriter.initialize(dbfPath, conf);
      dbfWriter.write(f);
    } finally {
      try {
        dbfWriter.close();
      } catch (Exception e) {
        e.printStackTrace();
      }
    }

    // Read the written file back and make sure it is similar to the original one
    try (DBFReader dbfReader = new DBFReader()) {
      dbfReader.initialize(new DataInputStream(new FileInputStream(dbfPath.toString())), new BeastOptions(conf));
      assertTrue("The output should have one record", dbfReader.nextKeyValue());
      IFeature f2 = dbfReader.getCurrentValue();
      for (int i : f.iNonGeomJ()) {
        // Compare the object or its string representation since the map type is converted to string when loaded
        assertTrue("Features attribute #"+i+" should be equal",
            f.get(i).equals(f2.get(i)) || f.get(i).toString().equals(f2.get(i).toString()));
      }
      assertFalse("The output should not have more than one record", dbfReader.nextKeyValue());
    }
  }

  public void testWriteNegativeIntegers() throws IOException {
    DBFWriter dbfWriter = new DBFWriter();
    Path dbfPath = new Path(scratchPath(), "output.dbf");
    Configuration conf = new Configuration();
    // Create a DBF file with one feature that has all the possible types
    Feature f = Feature.create(null, new String[]{"key1", "key2", "key3"},
        null, new Object[]{-123, (long) Integer.MIN_VALUE, Long.MAX_VALUE});
    try {
      dbfWriter.initialize(dbfPath, conf);
      dbfWriter.write(f);
    } finally {
      try {
        dbfWriter.close();
      } catch (Exception e) {
        e.printStackTrace();
      }
    }

    // Read the written file back and make sure it is similar to the original one
    try (DBFReader dbfReader = new DBFReader()) {
      dbfReader.initialize(new DataInputStream(new FileInputStream(dbfPath.toString())), new BeastOptions(conf));
      assertTrue("The output should have one record", dbfReader.nextKeyValue());
      IFeature f2 = dbfReader.getCurrentValue();
      assertEquals(f.getInt(1), f2.getInt(1));
      assertEquals(f.getLong(2), f2.getLong(2));
      assertEquals((double) f.getLong(3), f2.getDouble(3));
      assertFalse("The output should not have more than one record", dbfReader.nextKeyValue());
    }
  }

  public void testWriteNullAttributeTypes() throws IOException {
    DBFWriter dbfWriter = new DBFWriter();
    Path dbfPath = new Path(scratchPath(), "output.dbf");
    Configuration conf = new Configuration();
    // Create a DBF file with one feature that has all the possible types
    Feature f = Feature.create(null,
        new String[] {"key1", "key2", "key3", "key4", "key5", "key6"},
        new DataType[]{
            DataTypes.BooleanType, DataTypes.IntegerType, DataTypes.LongType,
            DataTypes.DoubleType, DataTypes.StringType, DataTypes.TimestampType
        }, null
    );
    try {
      dbfWriter.initialize(dbfPath, conf);
      dbfWriter.write(f);
    } finally {
      try {
        dbfWriter.close();
      } catch (Exception e) {

      }
    }

    // Read the written file back and make sure it is similar to the original one
    DBFReader dbfReader = new DBFReader();
    try {
      dbfReader.initialize(new DataInputStream(new FileInputStream(dbfPath.toString())), new BeastOptions(conf));
      assertTrue("The output should have one record", dbfReader.nextKeyValue());
      IFeature f2 = dbfReader.getCurrentValue();
      for (int i = 0; i < 6; i++)
        assertNull("Attribute " + i + " should be null", f2.get(i+1));
      assertEquals(DataTypes.BooleanType, f2.getDataType(1));
      assertEquals(DataTypes.IntegerType, f2.getDataType(2));
      assertEquals(DataTypes.IntegerType, f2.getDataType(3));
      assertEquals(DataTypes.IntegerType, f2.getDataType(4));
      assertEquals(DataTypes.StringType, f2.getDataType(5));
      assertEquals(DataTypes.TimestampType, f2.getDataType(6));
      assertFalse("The output should not have more than one record", dbfReader.nextKeyValue());
    } finally {
      dbfReader.close();
    }
  }

  public void testWriteUnicodeString() throws IOException {
    DBFWriter dbfWriter = new DBFWriter();
    Path dbfPath = new Path(scratchPath(), "output.dbf");
    Configuration conf = new Configuration();
    String word = "\u0633\u0644\u0627\u0645";
    // Create a DBF file with one feature that has all the possible types
    Feature f = Feature.create(null, new String[]{"key1"},
        null, new Object[]{word});
    try {
      dbfWriter.initialize(dbfPath, conf);
      dbfWriter.write(f);
    } finally {
      try {
        dbfWriter.close();
      } catch (Exception e) {
        e.printStackTrace();
      }
    }

    // Read the written file back and make sure it is similar to the original one
    try (DBFReader dbfReader = new DBFReader()) {
      dbfReader.initialize(new DataInputStream(new FileInputStream(dbfPath.toString())), new BeastOptions(conf));
      assertTrue("The output should have one record", dbfReader.nextKeyValue());
      IFeature f2 = dbfReader.getCurrentValue();
      assertEquals("Non-matching unicode characters "+ Arrays.toString(word.getBytes(StandardCharsets.UTF_8)) + " <-> "+
              Arrays.toString(f2.getString(1).getBytes(StandardCharsets.UTF_8)), word, f2.getString(1));
      assertFalse("The output should not have more than one record", dbfReader.nextKeyValue());
    }
  }

}

File: io/src/test/scala/edu/ucr/cs/bdlab/beast/io/shapefilev2/DBFWriterTest.scala
/*
 * Copyright 2020 University of California, Riverside
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package edu.ucr.cs.bdlab.beast.io.shapefilev2

import edu.ucr.cs.bdlab.test.BeastSpatialTest
import org.apache.commons.io.output.ByteArrayOutputStream
import org.apache.spark.sql.types._
import org.junit.runner.RunWith
import org.scalatest.FunSuite
import org.scalatest.junit.JUnitRunner

import java.io.{ByteArrayInputStream, DataInputStream}
import java.util.{Calendar, GregorianCalendar}

@RunWith(classOf[JUnitRunner])
class DBFWriterTest extends FunSuite with BeastSpatialTest {
  test("simple writer") {
    // Read a file and write it back
    val dbfIn = new DataInputStream(getClass.getResourceAsStream("/usa-major-cities/usa-major-cities.dbf"))
    val dbfFile = new DBFReader(dbfIn)
    val dbfOut = new ByteArrayOutputStream()
    val dbfWriter = new DBFWriter(dbfOut, dbfFile.schema)
    for (record <- dbfFile) {
      dbfWriter.write(record)
    }
    dbfWriter.close()
    dbfIn.close()

    // Read the file back and compare record-by-record
    val expectedOut = new DBFReader(new DataInputStream(getClass.getResourceAsStream("/usa-major-cities/usa-major-cities.dbf")))
    val actualOutput = new DBFReader(new DataInputStream(new ByteArrayInputStream(dbfOut.toByteArray)))
    try {
      // println(expectedOut.schema)
      // println(actualOutput.schema)
      var numRecords: Int = 0
      while (expectedOut.hasNext && actualOutput.hasNext) {
        val expectedRecord = expectedOut.next()
        val actualRecord = actualOutput.next()
        assertResult(actualRecord)(expectedRecord)
        numRecords += 1
      }
      assert(!expectedOut.hasNext, s"Read only $numRecords records but expected to read more")
      assert(!actualOutput.hasNext, s"Read all $numRecords records but some records are left")
    } finally {
      expectedOut.close()
      actualOutput.close()
    }

  }

  test("Write all attribute types") {
    // Create a DBF file with one feature that has all the possible types
    val date = new GregorianCalendar(DBFHelper.UTC)
    date.clear()
    date.set(2019, Calendar.OCTOBER, 9)
    val dateValue = new java.sql.Date(date.getTimeInMillis)
    val tags = Map("tag" -> "value")
    val schema = StructType(Seq(
      StructField("key1", BooleanType),
      StructField("key2", IntegerType),
      StructField("key3", ByteType),
      StructField("key4", ShortType),
      StructField("key5", LongType),
      StructField("key6", FloatType),
      StructField("key7", DoubleType),
      StructField("key8", StringType),
      StructField("key9", DateType),
      StructField("key10", MapType.apply(StringType, StringType)),
    ))
    val r: Array[Any] = Array[Any](true, 123, 12.toByte, 13.toShort, 6546543.toLong, 12.25f, 12.655847, "value9", dateValue, tags)
    val dbfOut = new ByteArrayOutputStream()
    val dbfWriter = new DBFWriter(dbfOut, schema)
    try {
      dbfWriter.write(r)
    } finally try dbfWriter.close()
    catch {
      case e: Exception => e.printStackTrace()
    }

    // Read the written file back and make sure it is similar to the original one
    val dbfReader = new DBFReader(new DataInputStream(new ByteArrayInputStream(dbfOut.toByteArray)))
    assert(dbfReader.hasNext)
    val r2 = dbfReader.next()
    for (i <- r.indices) {
      assert(r(i) == r2(i) || r(i).toString == r2(i).toString, s"Features attribute #$i should be equal")
    }
    assert(!dbfReader.hasNext, "The output should not have more than one record")
    dbfReader.close()
  }

  test("Write negative integers") {
    // Create a DBF file with one feature that has all the possible types
    val schema = StructType(Seq(
      StructField("key1", IntegerType),
      StructField("key2", LongType),
      StructField("key3", LongType),
    ))
    val r: Array[Any] = Array[Any](-123, Int.MinValue.toLong, Long.MaxValue)
    val dbfOut = new ByteArrayOutputStream()
    val dbfWriter = new DBFWriter(dbfOut, schema)
    try {
      dbfWriter.write(r)
    } finally try dbfWriter.close()
    catch {
      case e: Exception => e.printStackTrace()
    }

    // Read the written file back and make sure it is similar to the original one
    val dbfReader = new DBFReader(new DataInputStream(new ByteArrayInputStream(dbfOut.toByteArray)))
    assert(dbfReader.hasNext)
    val r2 = dbfReader.next()
    for (i <- r.indices) {
      assert(r(i) == r2(i) || r(i).toString == r2(i).toString, s"Features attribute #$i should be equal")
    }
    assert(!dbfReader.hasNext, "The output should not have more than one record")
    dbfReader.close()
  }

  test("Write null fields with known schema") {
    val schema = StructType(Seq(
      StructField("key1", BooleanType),
      StructField("key2", IntegerType),
      StructField("key3", LongType),
      StructField("key4", DoubleType),
      StructField("key5", StringType),
      StructField("key6", TimestampType),
    ))
    val dbfOut = new ByteArrayOutputStream()
    val dbfWriter = new DBFWriter(dbfOut, schema)
    val r = Array[Any](null, null, null, null, null, null)
    try {
      dbfWriter.write(r)
    } finally try dbfWriter.close()
    catch {
      case e: Exception => e.printStackTrace()
    }

    // Read it back
    val reader = new DBFReader(new DataInputStream(new ByteArrayInputStream(dbfOut.toByteArray)))
    assert(reader.hasNext, "The output should have one record")
    val r2 = reader.next()
    assert(r2.forall(_ == null), "All values should be null")
    val writtenSchema = reader.schema
    assertResult(BooleanType)(writtenSchema(0).dataType)
    assertResult(IntegerType)(writtenSchema(1).dataType)
    assertResult(IntegerType)(writtenSchema(2).dataType)
    assertResult(IntegerType)(writtenSchema(3).dataType)
    assertResult(StringType)(writtenSchema(4).dataType)
    assertResult(TimestampType)(writtenSchema(5).dataType)
    reader.close()
  }

  test("Write unicode string") {
    val schema = StructType(Seq(StructField("key1", StringType)))
    val dbfOut = new ByteArrayOutputStream()
    val dbfWriter = new DBFWriter(dbfOut, schema)
    val r = Array[Any]("\u0633\u0644\u0627\u0645")
    try {
      dbfWriter.write(r)
    } finally try dbfWriter.close()
    catch {
      case e: Exception => e.printStackTrace()
    }

    // Read it back
    val reader = new DBFReader(new DataInputStream(new ByteArrayInputStream(dbfOut.toByteArray)))
    assert(reader.hasNext, "The output should have one record")
    val r2 = reader.next()
    assertResult(r2(0))(r(0))
    reader.close()
  }
}
